feat: add `BitVec.(getMsbD, msb)_replicate, replicate_one`

luisacicolini opened the pull request:
This PR adds `BitVec.(getMsbD, msb)_replicate, replicate_one` theorems, corrects a non-terminal `simp` in `BitVec.getLsbD_replicate` and simplifies the proof of `BitVec.getElem_replicate` using the `cases` tactic.

Co-authored with @bollu.

tobiasgrosser commented:
@alexkeizer, can you do a round of reviews?

alexkeizer commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3208,6 +3208,84 @@ theorem getElem_replicate {n w : Nat} (x : BitVec w) (h : i < w * n) :
+    decide_eq_true_eq]
+    omega
+
+@[simp]
+private theorem mod_sub_eq_sub_mod {w n i : Nat} (hwn : i < w * n) (hn : 0 < n) :
```
A private simp theorem seems like a bad idea; can you remove this as a simp-lemma and use it explicitly where you need it?

Alternatively, put this as a Nat theorem where it belongs, so it doesn't have to be private

luisacicolini replied:
I've made it explicit for now, I am afraid it makes the proof much more confused, but let's see

alexkeizer replied:
How so? It should be as simple as simply adding `mod_sub_eq_sub_mod` to the `simp` invocation that was applying this one implicitly before?

That is, if you first squeeze all your simps (assuming this lemma is not used in a non-squeezed terminal simp), nothing should break if you then remove the simp annotation here

luisacicolini replied:
oh I see, i misinterpreted the explicitly and deleted the theorem, using its proof directly in `getMsbD_replicate` when necessary

luisacicolini replied:
but I'm now also refactoring the proof with `cases`, so I'll commit and update once that's done! thanks :)

alexkeizer commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3208,6 +3208,84 @@ theorem getElem_replicate {n w : Nat} (x : BitVec w) (h : i < w * n) :
+  induction n
+  case zero => omega
+  case succ n ih =>
+    simp_all [Nat.mul_add]
```
non-terminal simp

alexkeizer commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3208,6 +3208,84 @@ theorem getElem_replicate {n w : Nat} (x : BitVec w) (h : i < w * n) :
+          · rw [Nat.mul_comm]
+            omega
+          · rw [Nat.add_mul]
+            simp
```
non-terminal simp

alexkeizer commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3208,6 +3208,84 @@ theorem getElem_replicate {n w : Nat} (x : BitVec w) (h : i < w * n) :
+  case succ n ih =>
+    simp_all [Nat.mul_add]
+    by_cases h : i < w * n
+    · simp [show w * n + w - 1 -i = w + (w * n - 1 - i) by omega]
```
non-terminal simp

alexkeizer commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3208,6 +3208,84 @@ theorem getElem_replicate {n w : Nat} (x : BitVec w) (h : i < w * n) :
+          rw [← this]
+        omega
+      · omega
+
+
+@[simp]
```
```suggestion

@[simp]
```

alexkeizer commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3208,6 +3208,84 @@ theorem getElem_replicate {n w : Nat} (x : BitVec w) (h : i < w * n) :
+theorem getMsbD_replicate {n w : Nat} (x : BitVec w) :
+    (x.replicate n).getMsbD i =
+    (decide (i < w * n) && x.getMsbD (i % w)) := by
+  simp [getMsbD_eq_getLsbD]
```
non-terminal simp! (Please check this everywhere)

alexkeizer commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3208,6 +3208,84 @@ theorem getElem_replicate {n w : Nat} (x : BitVec w) (h : i < w * n) :
+    (x.replicate n).getMsbD i =
+    (decide (i < w * n) && x.getMsbD (i % w)) := by
+  simp [getMsbD_eq_getLsbD]
+  by_cases h₀ : 0 < w
+  · by_cases h₁ : i < w * n <;> by_cases h₂ : n = 0
```
I suspect the proof will be nicer if you use `cases w` and `cases n` rather than by_cases.

(`cases` is also morally the right thing to use, since by_cases invokes classical reasoning, which is wholly unnecessary here, but that is mostly my opinion on aesthetics that is not shared by the community as a whole)

alexkeizer commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3208,6 +3208,84 @@ theorem getElem_replicate {n w : Nat} (x : BitVec w) (h : i < w * n) :
+
+@[simp]
+theorem getMsbD_replicate {n w : Nat} (x : BitVec w) :
+    (x.replicate n).getMsbD i =
+    (decide (i < w * n) && x.getMsbD (i % w)) := by
```
Personally, when splitting in equality over two lines, I like to put the `=` on the start of the second line, so that's immediately obvious.
```suggestion
    (x.replicate n).getMsbD i
    = (decide (i < w * n) && x.getMsbD (i % w)) := by
```

This is not necessarily standard Lean style, though, so feel free to ignore

alexkeizer commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3208,6 +3208,84 @@ theorem getElem_replicate {n w : Nat} (x : BitVec w) (h : i < w * n) :
+    (x.replicate n).msb =
+    (decide (0 < n) && x.msb) := by
+  simp [BitVec.msb, getMsbD_replicate]
+  by_cases hn : 0 < n <;> by_cases hw : 0 < w
```
Again, these look like they should be regular cases
```suggestion
  cases n <;> cases w
```

(You might have to shuffle the rest of the proofs around a bit, I believe the different cases will be in a different order after this change)

luisacicolini commented:
thanks @alexkeizer for the review. I refactored the proof with `cases` and fixed all non terminal simps.

kim-em commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3208,6 +3208,88 @@ theorem getElem_replicate {n w : Nat} (x : BitVec w) (h : i < w * n) :
+@[simp]
+theorem replicate_append_replicate_eq {w n : Nat} {x : BitVec w} (h : w * (n + m) = w * n + w * m := by omega) :
+    x.replicate n ++ x.replicate m = (x.replicate (n + m)).cast h := by
+  apply BitVec.eq_of_getLsbD_eq
+  simp only [getLsbD_cast, getLsbD_replicate, getLsbD_append, getLsbD_replicate]
+  intros i
+  by_cases h₀ : i < w * m <;> by_cases h₁ : i < w * (n + m)
+  · simp only [h₀, decide_true, Bool.true_and, cond_true, h₁]
+  · rw [Nat.mul_add] at h₁
+    simp only [h₀, decide_true, Bool.true_and, cond_true, Bool.iff_and_self, decide_eq_true_eq]
+    omega
+  · have h₂ : i ≥ w * m := by omega
+    simp only [h₀, decide_false, Bool.false_and, show i - w * m < w * n by omega, decide_true,
+      Bool.true_and, cond_false, h₁]
+    congr 1
+    rw [Nat.sub_mul_mod (by omega)]
+  · simp only [h₀, decide_false, Bool.false_and, cond_false, h₁, Bool.and_eq_false_imp,
+    decide_eq_true_eq]
+    omega
```
Merge master and use
```
  ext i h₁
  simp only [getLsbD_append]
  split <;> rename_i h₂
  · simp [h₂]
    omega
  · have h₃ : i - w * m < w * n := by omega
    simp [h, h₁, h₃, Nat.sub_mul_mod (by simpa using h₂)]
```

kim-em commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3208,6 +3208,88 @@ theorem getElem_replicate {n w : Nat} (x : BitVec w) (h : i < w * n) :
+      · omega
+
+@[simp]
+theorem getMsbD_replicate {n w : Nat} (x : BitVec w) :
+    (x.replicate n).getMsbD i
+    = (decide (i < w * n) && x.getMsbD (i % w)) := by
```
I think this would be much simpler if you introduced `BitVec.reverse` first, and prove this via 
`rw [← getLsbD_reverse, reverse_replicate, getLsbD_replicate, getLsbD_reverse]`

Could you have a go at that? It should avoid all the index munging.

luisacicolini replied:
thanks @kim-em! Am currently working on the `reverse` on a different branch to avoiding adding too much stuff here (I am trying to follow the same design strategies as `List`, treating the bitvec as a list of bools).

kim-em commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3208,6 +3208,88 @@ theorem getElem_replicate {n w : Nat} (x : BitVec w) (h : i < w * n) :
+    decide_eq_true_eq]
+    omega
+
+theorem mod_sub_eq_sub_mod {w n i : Nat} (hwn : i < w * n) (hn : 0 < n) :
+    (w * n - 1 - i) % w = w - 1 - i % w := by
```
This proof can probably be shortened! But let's see if we can avoid it entirely.

luisacicolini commented:
After the merging of #6476 I updated the proof of `getMsbD_replicate` and updated the description (some theorems seem new in the diff, but it's because I had to move them under the `*_reverse` theorems).

luisacicolini commented:
changelog-library

tobiasgrosser commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3796,6 +3772,47 @@ theorem reverse_replicate {n : Nat} {x : BitVec w} :
+
+@[simp]
+theorem msb_replicate {n w : Nat} (x : BitVec w) :
+    (x.replicate n).msb =
+    (decide (0 < n) && x.msb) := by
```
```suggestion
    (x.replicate n).msb = (decide (0 < n) && x.msb) := by
```

alexkeizer replied:
@luisacicolini you resolved this, but did not seem to apply the suggestion? I'd second Tobias' version

tobiasgrosser commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3796,6 +3772,47 @@ theorem reverse_replicate {n : Nat} {x : BitVec w} :
+theorem msb_replicate {n w : Nat} (x : BitVec w) :
+    (x.replicate n).msb =
+    (decide (0 < n) && x.msb) := by
+  simp [BitVec.msb, getMsbD_replicate]
```
Non terminal simp.

tobiasgrosser commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3796,6 +3772,47 @@ theorem reverse_replicate {n : Nat} {x : BitVec w} :
+
+@[simp]
+theorem getMsbD_replicate {n w : Nat} (x : BitVec w) :
+    (x.replicate n).getMsbD i =
+    (decide (i < w * n) && x.getMsbD (i % w)) := by
```
```suggestion
    (x.replicate n).getMsbD i = (decide (i < w * n) && x.getMsbD (i % w)) := by
```

tobiasgrosser commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3813,6 +3720,99 @@ theorem toInt_abs_eq_natAbs_of_ne_intMin {x : BitVec w} (hx : x ≠ intMin w) :
+    simp [reverse_append, ih]
+
+@[simp]
+theorem getLsbD_replicate {n w : Nat} (x : BitVec w) :
+    (x.replicate n).getLsbD i =
```
Should we not use `{}` around `x : BitVec w`? There are several other cases like this.

tobiasgrosser commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3813,6 +3720,99 @@ theorem toInt_abs_eq_natAbs_of_ne_intMin {x : BitVec w} (hx : x ≠ intMin w) :
+@[simp]
+theorem getLsbD_replicate {n w : Nat} (x : BitVec w) :
+    (x.replicate n).getLsbD i =
+    (decide (i < w * n) && x.getLsbD (i % w)) := by
```
Maybe put the equality in a single line.

tobiasgrosser commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3078,8 +3078,8 @@ theorem getMsbD_rotateLeft_of_lt {n w : Nat} {x : BitVec w} (hi : r < w):
         have h₂ : (r + n) < 2 * (w + 1) := by omega
+        rw [← Nat.sub_mul_eq_mod_of_lt_of_le (i := (r + n)) (w := (w + 1)) (n := 1) (by omega) (by omega)]
         congr 1
-        rw [← Nat.sub_mul_eq_mod_of_lt_of_le (n := 1) (by omega) (by omega), Nat.mul_one]
```
Is this change needed?

tobiasgrosser replied:
There is still a change in this line. Do we need this change at all? It does not really seem to make anything simpler and adds noise that can be avoided, no?

luisacicolini replied:
I now simplified it further and removed the unnecessary specification. theorem `Nat.mul_one` is not necessary here, but if this change is too much happy to revert it.

alexkeizer replied:
I think the current version is fine

tobiasgrosser commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3408,30 +3408,6 @@ theorem replicate_succ {x : BitVec w} :
-    (x.replicate n)[i] = if h' : w = 0 then false else x[i % w]'(@Nat.mod_lt i w (by omega)) := by
-  simp only [← getLsbD_eq_getElem, getLsbD_replicate]
-  by_cases h' : w = 0 <;> simp [h'] <;> omega
-
```
Maybe add to the PR message info why these theorems have been moved.

alexkeizer commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3796,6 +3772,45 @@ theorem reverse_replicate {n : Nat} {x : BitVec w} :
+
+@[simp]
+theorem replicate_one {w : Nat} {x : BitVec w} (h : w = w * 1 := by omega) :
+    (x.replicate 1) = x.cast h := by simp [replicate, h]
```
```suggestion
    (x.replicate 1) = x.cast h := by 
  simp [replicate, h]
```
I generally don't like having a proof on the same line as the statement, unless it's a `rfl` proof or the whole theorem fits into one line

tobiasgrosser commented:
@luisacicolini, the PR title still needs updating.

alexkeizer commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3796,6 +3795,22 @@ theorem reverse_replicate {n : Nat} {x : BitVec w} :
+@[simp]
+theorem replicate_one {w : Nat} {x : BitVec w} (h : w = w * 1 := by rw [Nat.mul_one]) :
+    (x.replicate 1) = x.cast h := by
+  simp [replicate, h]
```
I was a bit surprised this needed a cast, but [https://live.lean-lang.org/#codez=FAUwHghgtgDgNiABACjIgXIgchALgSg0TQCpEBGRAXmIxoCMBPRYRRAJwDM5hRJYEKNJhwEilMmhrCGzVh268AxAGMAFiBUBrbHgB0UAK5wgA](it seems) that indeed `w * 1` is not defeq to `w`, so this looks good

alexkeizer commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3402,14 +3402,19 @@ theorem and_one_eq_setWidth_ofBool_getLsbD {x : BitVec w} :
   simp [replicate]
 
+@[simp]
+theorem replicate_one {w : Nat} {x : BitVec w} (h : w = w * 1 := Eq.symm (Nat.mul_one _) ) :
```
```suggestion
theorem replicate_one {w : Nat} {x : BitVec w} (h : w = w * 1 := Eq.symm (Nat.mul_one _)) :
```

luisacicolini commented:
awaiting-review

kim-em commented on src/Init/Data/BitVec/Lemmas.lean:
```diff
@@ -3402,14 +3402,19 @@ theorem and_one_eq_setWidth_ofBool_getLsbD {x : BitVec w} :
   simp [replicate]
 
+@[simp]
+theorem replicate_one {w : Nat} {x : BitVec w} (h : w = w * 1 := Eq.symm (Nat.mul_one _)) :
```
I don't understand why `h` is an argument here. Just inline the default value into the argument of the cast. Proof irrelevance ensures that `simp` and `rw` can still apply the lemma even if a "different" proof appears.

luisacicolini commented:
awaiting-review

bollu reviewed (approved):
LGTM
